sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -699,6 +700,28 @@ case class UnionExec(children: Seq[SparkPlan]) extends SparkPlan {
     }
   }
 
+  private lazy val childrenRDDs = children.map(_.execute())
+
+  override def outputPartitioning: Partitioning = {
+    try {
+      val nonEmptyRdds = childrenRDDs.filter(!_.partitions.isEmpty)
+      if (sparkContext.isPartitionerAwareUnion(nonEmptyRdds)) {
+        // `isPartitionerAwareUnion` ensures that at least one child is non-empty.
+        children.head.outputPartitioning
+      } else {
+        super.outputPartitioning
+      }
+    } catch {
+      // If any child operator doesn't support `execute`, we cannot determine the
+      // partitioning. Even if it is other exception, we also simply fall back to
+      // the default partitioning. Note that for such cases, it means that these
+      // child operator will be replaced by Spark in query planning later, in other
+      // words, `execute` won't be actually called on them during the execution of
+      // this plan. So we can safely return the default partitioning.
+      case e if NonFatal(e) => super.outputPartitioning
```

Comment:
This handles nodes that don't implement `execute` method. The reason is described like the comment said.